File: src/main/scala/ff/compiler/Syntax.scala
package ff.compiler
import ff.core.Array_._

import ff.core.ArrayBuilder_._

import ff.core.Bool_._

import ff.core.Char_._

import ff.core.Core_._

import ff.core.FileSystem_._

import ff.core.Int_._

import ff.core.List_._

import ff.core.Log_._

import ff.core.Map_._

import ff.core.Nothing_._

import ff.core.Option_._

import ff.core.Pair_._

import ff.core.Set_._

import ff.core.String_._

import ff.core.System_._

import ff.core.Try_._

import ff.core.Unit_._
object Syntax_ {

case class Location(file_ : ff.core.String_.String, line_ : ff.core.Int_.Int, column_ : ff.core.Int_.Int)

case class CompileError(at_ : ff.compiler.Syntax_.Location, message_ : ff.core.String_.String)

case class Module(packagePair_ : ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String], file_ : ff.core.String_.String, dependencies_ : ff.core.List_.List[ff.compiler.Syntax_.DDependency], imports_ : ff.core.List_.List[ff.compiler.Syntax_.DImport], types_ : ff.core.List_.List[ff.compiler.Syntax_.DType], traits_ : ff.core.List_.List[ff.compiler.Syntax_.DTrait], instances_ : ff.core.List_.List[ff.compiler.Syntax_.DInstance], extends_ : ff.core.List_.List[ff.compiler.Syntax_.DExtend], lets_ : ff.core.List_.List[ff.compiler.Syntax_.DLet], functions_ : ff.core.List_.List[ff.compiler.Syntax_.DFunction])

case class DDependency(at_ : ff.compiler.Syntax_.Location, package_ : ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String], safety_ : ff.compiler.Syntax_.Safety, goodVersions_ : ff.core.List_.List[ff.compiler.Syntax_.Version], badVersions_ : ff.core.List_.List[ff.compiler.Syntax_.Version])

case class DImport(at_ : ff.compiler.Syntax_.Location, alias_ : ff.core.String_.String, package_ : ff.core.Pair_.Pair[ff.core.String_.String, ff.core.String_.String], directory_ : ff.core.List_.List[ff.core.String_.String], file_ : ff.core.String_.String)

case class DFunction(at_ : ff.compiler.Syntax_.Location, signature_ : ff.compiler.Syntax_.Signature, body_ : ff.compiler.Syntax_.Lambda, targets_ : ff.compiler.Syntax_.Targets)

case class DLet(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, variableType_ : ff.compiler.Syntax_.Type, value_ : ff.compiler.Syntax_.Term, targets_ : ff.compiler.Syntax_.Targets)

case class DExtend(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, generics_ : ff.core.List_.List[ff.core.String_.String], constraints_ : ff.core.List_.List[ff.compiler.Syntax_.Constraint], type_ : ff.compiler.Syntax_.Type, methods_ : ff.core.List_.List[ff.compiler.Syntax_.DFunction])

case class DType(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, generics_ : ff.core.List_.List[ff.core.String_.String], constraints_ : ff.core.List_.List[ff.compiler.Syntax_.Constraint], commonFields_ : ff.core.List_.List[ff.compiler.Syntax_.Parameter], variants_ : ff.core.List_.List[ff.compiler.Syntax_.Variant], targets_ : ff.compiler.Syntax_.Targets)

case class DTrait(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, generics_ : ff.core.List_.List[ff.core.String_.String], constraints_ : ff.core.List_.List[ff.compiler.Syntax_.Constraint], generatorParameters_ : ff.core.List_.List[ff.compiler.Syntax_.Parameter], methods_ : ff.core.List_.List[ff.compiler.Syntax_.Signature], methodDefaults_ : ff.core.List_.List[ff.core.Pair_.Pair[ff.core.String_.String, ff.compiler.Syntax_.Lambda]], methodGenerators_ : ff.core.List_.List[ff.core.Pair_.Pair[ff.core.String_.String, ff.compiler.Syntax_.Lambda]])

case class DInstance(at_ : ff.compiler.Syntax_.Location, generics_ : ff.core.List_.List[ff.core.String_.String], constraints_ : ff.core.List_.List[ff.compiler.Syntax_.Constraint], traitType_ : ff.compiler.Syntax_.Type, generatorArguments_ : ff.core.List_.List[ff.compiler.Syntax_.Argument], methods_ : ff.core.List_.List[ff.compiler.Syntax_.DFunction])

sealed abstract class Term extends Product with Serializable {
    val at_ : ff.compiler.Syntax_.Location
}
case class EString(at_ : ff.compiler.Syntax_.Location, value_ : ff.core.String_.String) extends Term
case class EChar(at_ : ff.compiler.Syntax_.Location, value_ : ff.core.String_.String) extends Term
case class EInt(at_ : ff.compiler.Syntax_.Location, value_ : ff.core.String_.String) extends Term
case class EFloat(at_ : ff.compiler.Syntax_.Location, value_ : ff.core.String_.String) extends Term
case class EVariable(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, generics_ : ff.core.List_.List[ff.compiler.Syntax_.Type], instances_ : ff.core.List_.List[ff.compiler.Syntax_.Instance]) extends Term
case class ELambda(at_ : ff.compiler.Syntax_.Location, lambda_ : ff.compiler.Syntax_.Lambda) extends Term
case class EFunctions(at_ : ff.compiler.Syntax_.Location, functions_ : ff.core.List_.List[ff.compiler.Syntax_.DFunction], body_ : ff.compiler.Syntax_.Term) extends Term
case class ELet(at_ : ff.compiler.Syntax_.Location, mutable_ : ff.core.Bool_.Bool, name_ : ff.core.String_.String, valueType_ : ff.compiler.Syntax_.Type, value_ : ff.compiler.Syntax_.Term, body_ : ff.compiler.Syntax_.Term) extends Term
case class ESequential(at_ : ff.compiler.Syntax_.Location, before_ : ff.compiler.Syntax_.Term, after_ : ff.compiler.Syntax_.Term) extends Term
case class EAssign(at_ : ff.compiler.Syntax_.Location, operator_ : ff.core.String_.String, variable_ : ff.core.String_.String, value_ : ff.compiler.Syntax_.Term) extends Term
case class EAssignField(at_ : ff.compiler.Syntax_.Location, operator_ : ff.core.String_.String, record_ : ff.compiler.Syntax_.Term, field_ : ff.core.String_.String, value_ : ff.compiler.Syntax_.Term) extends Term
case class EPipe(at_ : ff.compiler.Syntax_.Location, value_ : ff.compiler.Syntax_.Term, function_ : ff.compiler.Syntax_.Term) extends Term
case class ECall(at_ : ff.compiler.Syntax_.Location, tailCall_ : ff.core.Bool_.Bool, function_ : ff.compiler.Syntax_.Term, typeArguments_ : ff.core.List_.List[ff.compiler.Syntax_.Type], arguments_ : ff.core.List_.List[ff.compiler.Syntax_.Argument]) extends Term
case class EList(at_ : ff.compiler.Syntax_.Location, elementType_ : ff.compiler.Syntax_.Type, items_ : ff.core.List_.List[ff.core.Pair_.Pair[ff.compiler.Syntax_.Term, ff.core.Bool_.Bool]]) extends Term
case class ECopy(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, record_ : ff.compiler.Syntax_.Term, arguments_ : ff.core.List_.List[ff.compiler.Syntax_.Field]) extends Term
case class EVariant(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, typeArguments_ : ff.core.List_.List[ff.compiler.Syntax_.Type], arguments_ : ff.core.Option_.Option[ff.core.List_.List[ff.compiler.Syntax_.Argument]]) extends Term
case class EVariantIs(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, typeArguments_ : ff.core.List_.List[ff.compiler.Syntax_.Type]) extends Term
case class ERecord(at_ : ff.compiler.Syntax_.Location, fields_ : ff.core.List_.List[ff.compiler.Syntax_.Field]) extends Term
case class EField(at_ : ff.compiler.Syntax_.Location, record_ : ff.compiler.Syntax_.Term, field_ : ff.core.String_.String) extends Term
case class EWildcard(at_ : ff.compiler.Syntax_.Location, index_ : ff.core.Int_.Int) extends Term

case class MatchCase(at_ : ff.compiler.Syntax_.Location, patterns_ : ff.core.List_.List[ff.compiler.Syntax_.MatchPattern], condition_ : ff.core.Option_.Option[ff.compiler.Syntax_.Term], body_ : ff.compiler.Syntax_.Term)

sealed abstract class MatchPattern extends Product with Serializable {
    val at_ : ff.compiler.Syntax_.Location
}
case class PVariable(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.Option_.Option[ff.core.String_.String]) extends MatchPattern
case class PVariant(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, patterns_ : ff.core.List_.List[ff.compiler.Syntax_.MatchPattern]) extends MatchPattern
case class PVariantAs(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, variable_ : ff.core.Option_.Option[ff.core.String_.String]) extends MatchPattern
case class PAlias(at_ : ff.compiler.Syntax_.Location, pattern_ : ff.compiler.Syntax_.MatchPattern, variable_ : ff.core.String_.String) extends MatchPattern
case class PList(at_ : ff.compiler.Syntax_.Location, itemType_ : ff.compiler.Syntax_.Type, items_ : ff.core.List_.List[ff.core.Pair_.Pair[ff.compiler.Syntax_.MatchPattern, ff.core.Bool_.Bool]]) extends MatchPattern

case class Signature(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, generics_ : ff.core.List_.List[ff.core.String_.String], constraints_ : ff.core.List_.List[ff.compiler.Syntax_.Constraint], parameters_ : ff.core.List_.List[ff.compiler.Syntax_.Parameter], returnType_ : ff.compiler.Syntax_.Type)

case class Lambda(at_ : ff.compiler.Syntax_.Location, cases_ : ff.core.List_.List[ff.compiler.Syntax_.MatchCase])

case class Variant(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, fields_ : ff.core.List_.List[ff.compiler.Syntax_.Parameter], targets_ : ff.compiler.Syntax_.Targets)

case class Parameter(at_ : ff.compiler.Syntax_.Location, mutable_ : ff.core.Bool_.Bool, name_ : ff.core.String_.String, valueType_ : ff.compiler.Syntax_.Type, default_ : ff.core.Option_.Option[ff.compiler.Syntax_.Term])

case class Argument(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.Option_.Option[ff.core.String_.String], value_ : ff.compiler.Syntax_.Term)

case class Field(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, value_ : ff.compiler.Syntax_.Term)

case class Constraint(representation_ : ff.compiler.Syntax_.Type)

case class Targets(scala_ : ff.core.Option_.Option[ff.core.String_.String], javaScript_ : ff.core.Option_.Option[ff.core.String_.String])

sealed abstract class Type extends Product with Serializable {
    val at_ : ff.compiler.Syntax_.Location
}
case class TConstructor(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, generics_ : ff.core.List_.List[ff.compiler.Syntax_.Type]) extends Type
case class TVariable(at_ : ff.compiler.Syntax_.Location, index_ : ff.core.Int_.Int) extends Type

sealed abstract class Instance extends Product with Serializable {
    val at_ : ff.compiler.Syntax_.Location
}
case class IConstructor(at_ : ff.compiler.Syntax_.Location, name_ : ff.core.String_.String, arguments_ : ff.core.List_.List[ff.compiler.Syntax_.Instance]) extends Instance
case class IField(at_ : ff.compiler.Syntax_.Location, instance_ : ff.compiler.Syntax_.Instance, field_ : ff.core.String_.String) extends Instance
case class IVariable(at_ : ff.compiler.Syntax_.Location, index_ : ff.core.Int_.Int) extends Instance

sealed abstract class Safety extends Product with Serializable
case class Safe() extends Safety
case class Unsafe() extends Safety
case class Trust() extends Safety

case class Version(at_ : ff.compiler.Syntax_.Location, major_ : ff.core.Int_.Int, minor_ : ff.core.Int_.Int, patch_ : ff.core.Int_.Int)


def Location_show(self_ : ff.compiler.Syntax_.Location) : ff.core.String_.String = (self_) match {
case (self_) =>
(((((("in " + self_.file_) + " ") + "at line ") + self_.line_) + ", column ") + self_.column_)
}

def Type_show(self_ : ff.compiler.Syntax_.Type) : ff.core.String_.String = (self_) match {
case (self_) =>
pipe_dot(self_)({
case (ff.compiler.Syntax_.TConstructor(at_, name_, generics_)) =>
ff.core.Option_.Option_else(self_ = ff.core.Core_.if_[ff.core.String_.String](condition_ = ff.core.List_.List_getEmpty[ff.compiler.Syntax_.Type](self_ = generics_), body_ = {() =>
name_
}), body_ = {() =>
(((name_ + "[") + ff.core.List_.List_join(self_ = ff.core.List_.List_map[ff.compiler.Syntax_.Type, ff.core.String_.String](self_ = generics_, body_ = {(_w1) =>
ff.compiler.Syntax_.Type_show(self_ = _w1)
}), separator_ = ", ")) + "]")
})
case (ff.compiler.Syntax_.TVariable(at_, index_)) =>
("$" + index_)
})
}


}

File: src/main/js/ff/compiler/Syntax.mjs
import * as ff_compiler_Syntax from "../../ff/compiler/Syntax.mjs"

import * as ff_core_Array from "../../ff/core/Array.mjs"

import * as ff_core_ArrayBuilder from "../../ff/core/ArrayBuilder.mjs"

import * as ff_core_Bool from "../../ff/core/Bool.mjs"

import * as ff_core_Char from "../../ff/core/Char.mjs"

import * as ff_core_Core from "../../ff/core/Core.mjs"

import * as ff_core_FileSystem from "../../ff/core/FileSystem.mjs"

import * as ff_core_Int from "../../ff/core/Int.mjs"

import * as ff_core_List from "../../ff/core/List.mjs"

import * as ff_core_Log from "../../ff/core/Log.mjs"

import * as ff_core_Map from "../../ff/core/Map.mjs"

import * as ff_core_Nothing from "../../ff/core/Nothing.mjs"

import * as ff_core_Option from "../../ff/core/Option.mjs"

import * as ff_core_Pair from "../../ff/core/Pair.mjs"

import * as ff_core_Set from "../../ff/core/Set.mjs"

import * as ff_core_String from "../../ff/core/String.mjs"

import * as ff_core_System from "../../ff/core/System.mjs"

import * as ff_core_Try from "../../ff/core/Try.mjs"

import * as ff_core_Unit from "../../ff/core/Unit.mjs"

// type Location
export function Location(file_, line_, column_) {
return {_: 'Location', file_, line_, column_};
}

// type CompileError
export function CompileError(at_, message_) {
return {_: 'CompileError', at_, message_};
}

// type Module
export function Module(packagePair_, file_, dependencies_, imports_, types_, traits_, instances_, extends_, lets_, functions_) {
return {_: 'Module', packagePair_, file_, dependencies_, imports_, types_, traits_, instances_, extends_, lets_, functions_};
}

// type DDependency
export function DDependency(at_, package_, safety_, goodVersions_, badVersions_) {
return {_: 'DDependency', at_, package_, safety_, goodVersions_, badVersions_};
}

// type DImport
export function DImport(at_, alias_, package_, directory_, file_) {
return {_: 'DImport', at_, alias_, package_, directory_, file_};
}

// type DFunction
export function DFunction(at_, signature_, body_, targets_) {
return {_: 'DFunction', at_, signature_, body_, targets_};
}

// type DLet
export function DLet(at_, name_, variableType_, value_, targets_) {
return {_: 'DLet', at_, name_, variableType_, value_, targets_};
}

// type DExtend
export function DExtend(at_, name_, generics_, constraints_, type_, methods_) {
return {_: 'DExtend', at_, name_, generics_, constraints_, type_, methods_};
}

// type DType
export function DType(at_, name_, generics_, constraints_, commonFields_, variants_, targets_) {
return {_: 'DType', at_, name_, generics_, constraints_, commonFields_, variants_, targets_};
}

// type DTrait
export function DTrait(at_, name_, generics_, constraints_, generatorParameters_, methods_, methodDefaults_, methodGenerators_) {
return {_: 'DTrait', at_, name_, generics_, constraints_, generatorParameters_, methods_, methodDefaults_, methodGenerators_};
}

// type DInstance
export function DInstance(at_, generics_, constraints_, traitType_, generatorArguments_, methods_) {
return {_: 'DInstance', at_, generics_, constraints_, traitType_, generatorArguments_, methods_};
}

// type Term
export function EString(at_, value_) {
return {_: 'EString', at_, value_};
}
export function EChar(at_, value_) {
return {_: 'EChar', at_, value_};
}
export function EInt(at_, value_) {
return {_: 'EInt', at_, value_};
}
export function EFloat(at_, value_) {
return {_: 'EFloat', at_, value_};
}
export function EVariable(at_, name_, generics_, instances_) {
return {_: 'EVariable', at_, name_, generics_, instances_};
}
export function ELambda(at_, lambda_) {
return {_: 'ELambda', at_, lambda_};
}
export function EFunctions(at_, functions_, body_) {
return {_: 'EFunctions', at_, functions_, body_};
}
export function ELet(at_, mutable_, name_, valueType_, value_, body_) {
return {_: 'ELet', at_, mutable_, name_, valueType_, value_, body_};
}
export function ESequential(at_, before_, after_) {
return {_: 'ESequential', at_, before_, after_};
}
export function EAssign(at_, operator_, variable_, value_) {
return {_: 'EAssign', at_, operator_, variable_, value_};
}
export function EAssignField(at_, operator_, record_, field_, value_) {
return {_: 'EAssignField', at_, operator_, record_, field_, value_};
}
export function EPipe(at_, value_, function_) {
return {_: 'EPipe', at_, value_, function_};
}
export function ECall(at_, tailCall_, function_, typeArguments_, arguments_) {
return {_: 'ECall', at_, tailCall_, function_, typeArguments_, arguments_};
}
export function EList(at_, elementType_, items_) {
return {_: 'EList', at_, elementType_, items_};
}
export function ECopy(at_, name_, record_, arguments_) {
return {_: 'ECopy', at_, name_, record_, arguments_};
}
export function EVariant(at_, name_, typeArguments_, arguments_) {
return {_: 'EVariant', at_, name_, typeArguments_, arguments_};
}
export function EVariantIs(at_, name_, typeArguments_) {
return {_: 'EVariantIs', at_, name_, typeArguments_};
}
export function ERecord(at_, fields_) {
return {_: 'ERecord', at_, fields_};
}
export function EField(at_, record_, field_) {
return {_: 'EField', at_, record_, field_};
}
export function EWildcard(at_, index_) {
return {_: 'EWildcard', at_, index_};
}

// type MatchCase
export function MatchCase(at_, patterns_, condition_, body_) {
return {_: 'MatchCase', at_, patterns_, condition_, body_};
}

// type MatchPattern
export function PVariable(at_, name_) {
return {_: 'PVariable', at_, name_};
}
export function PVariant(at_, name_, patterns_) {
return {_: 'PVariant', at_, name_, patterns_};
}
export function PVariantAs(at_, name_, variable_) {
return {_: 'PVariantAs', at_, name_, variable_};
}
export function PAlias(at_, pattern_, variable_) {
return {_: 'PAlias', at_, pattern_, variable_};
}
export function PList(at_, itemType_, items_) {
return {_: 'PList', at_, itemType_, items_};
}

// type Signature
export function Signature(at_, name_, generics_, constraints_, parameters_, returnType_) {
return {_: 'Signature', at_, name_, generics_, constraints_, parameters_, returnType_};
}

// type Lambda
export function Lambda(at_, cases_) {
return {_: 'Lambda', at_, cases_};
}

// type Variant
export function Variant(at_, name_, fields_, targets_) {
return {_: 'Variant', at_, name_, fields_, targets_};
}

// type Parameter
export function Parameter(at_, mutable_, name_, valueType_, default_) {
return {_: 'Parameter', at_, mutable_, name_, valueType_, default_};
}

// type Argument
export function Argument(at_, name_, value_) {
return {_: 'Argument', at_, name_, value_};
}

// type Field
export function Field(at_, name_, value_) {
return {_: 'Field', at_, name_, value_};
}

// type Constraint
export function Constraint(representation_) {
return {_: 'Constraint', representation_};
}

// type Targets
export function Targets(scala_, javaScript_) {
return {_: 'Targets', scala_, javaScript_};
}

// type Type
export function TConstructor(at_, name_, generics_) {
return {_: 'TConstructor', at_, name_, generics_};
}
export function TVariable(at_, index_) {
return {_: 'TVariable', at_, index_};
}

// type Instance
export function IConstructor(at_, name_, arguments_) {
return {_: 'IConstructor', at_, name_, arguments_};
}
export function IField(at_, instance_, field_) {
return {_: 'IField', at_, instance_, field_};
}
export function IVariable(at_, index_) {
return {_: 'IVariable', at_, index_};
}

// type Safety
const Safe$ = {_: 'Safe'};
export function Safe() {
return Safe$;
}
const Unsafe$ = {_: 'Unsafe'};
export function Unsafe() {
return Unsafe$;
}
const Trust$ = {_: 'Trust'};
export function Trust() {
return Trust$;
}

// type Version
export function Version(at_, major_, minor_, patch_) {
return {_: 'Version', at_, major_, minor_, patch_};
}





export function Location_show(self_) {
return (((((("in " + self_.file_) + " ") + "at line ") + self_.line_) + ", column ") + self_.column_)
}

export function Type_show(self_) {
return (((_1) => {
{
if(_1._ === 'TConstructor') {
const at_ = _1.at_
const name_ = _1.name_
const generics_ = _1.generics_
if(ff_core_List.List_isEmpty(generics_)) {
return name_
} else {
return (((name_ + "[") + ff_core_List.List_join(ff_core_List.List_map(generics_, ((_w1) => {
return ff_compiler_Syntax.Type_show(_w1)
})), ", ")) + "]")
}
return
}
}
{
if(_1._ === 'TVariable') {
const at_ = _1.at_
const index_ = _1.index_
return ("$" + index_)
return
}
}
throw new Error('Unexhaustive pattern match')
}))(self_)
}


